File: src/NUISingleChoice.ts
import type { SingleChoiceButton, SingleChoiceOptions } from './Alert';
import NativeAlert from './NativeAlert';

type NUISingleChoiceOptions = Omit<
  SingleChoiceOptions,
  'positive' | 'neutral' | 'negative'
> & {
  title: string;
  items: Array<string>;
  positiveText?: string | undefined;
  negativeText?: string | undefined;
  neutralText?: string | undefined;
  defaultSelectedIndex?: number | undefined;
};

export class NUISingleChoice {
  private allOptions: NUISingleChoiceOptions;

  constructor(
    title: string,
    private items?: Array<SingleChoiceButton>,
    private options: SingleChoiceOptions = {}
  ) {
    this.allOptions = {
      title: title,
      items:
        this.items?.filter((item) => !!item.text).map((item) => item.text) ||
        [],
      ...(options || {}),
      positiveText:
        options.positive?.text ||
        (missingNegativeAndNeutral(options) ? 'OK' : undefined),
      negativeText: options.negative?.text,
      neutralText: options.neutral?.text,
      defaultSelectedIndex: this.items?.findIndex(
        (item) => item.defaultSelected
      ),
    };
  }

  show() {
    NativeAlert.singleChoice(
      this.allOptions,
      this.callback.bind(this),
      this.error.bind(this)
    );
  }

  private callback(pressed: string, selectedIndex: string) {
    const selectedValue = this.items?.[parseInt(selectedIndex, 10)];

    if (pressed === 'dismissed') {
      this.options.onDismiss?.(selectedValue?.value);
    } else if (pressed === 'positive') {
      this.options.positive?.onPress?.(selectedValue?.value);
    } else if (pressed === 'negative') {
      this.options.negative?.onPress?.(selectedValue?.value);
    } else if (pressed === 'neutral') {
      this.options.neutral?.onPress?.(selectedValue?.value);
    }
  }

  private error() {}
}

function missingNegativeAndNeutral(options: SingleChoiceOptions): boolean {
  return !options.negative && !options.neutral;
}

File: android/src/main/java/com/nui/alert/NUISingleChoice.kt
package com.nui.alert

import android.app.Dialog
import android.content.Context
import android.content.DialogInterface
import android.os.Bundle
import androidx.fragment.app.DialogFragment
import com.facebook.react.bridge.Callback
import com.facebook.react.bridge.ReactApplicationContext
import com.facebook.react.bridge.ReadableMap
import com.google.android.material.dialog.MaterialAlertDialogBuilder

class NUISingleChoice : DialogFragment, DialogInterface.OnClickListener {
  private val listener: NUISingleChoiceListener?
  private val options: SingleChoiceOptions?
  private var selectedIndex: Int? = null

  constructor() {
    listener = null
    options = null
  }

  internal constructor(arguments: ReadableMap, callback: Callback, reactApplicationContext: ReactApplicationContext) {
    this.listener = NUISingleChoiceListener(callback, reactApplicationContext)
    this.options = SingleChoiceOptions(arguments)
  }

  override fun onCreateDialog(savedInstanceState: Bundle?): Dialog =
    createDialog(requireActivity(), this)

  override fun onDismiss(dialog: DialogInterface) {
    super.onDismiss(dialog)
    listener?.onDismiss(dialog)
  }

  override fun onClick(dialog: DialogInterface, which: Int) {
    listener?.onClick(dialog, which)
  }

  private fun createDialog(
    activityContext: Context,
    fragment: DialogInterface.OnClickListener
  ): Dialog {
    this.isCancelable = options?.isCancelable ?: false

    val builder = MaterialAlertDialogBuilder(activityContext, options!!.headerAlignmentStyle)

    if (options.title.isNotEmpty()) {
      builder.setTitle(options.title)
    }

    if (!options.icon.isNullOrEmpty()) {
      builder.setIcon(IconFinder(requireContext()).getId(options.icon!!))
    }

    if (options.positiveText.isNotEmpty()) {
      builder.setPositiveButton(options.positiveText, listener)
    }

    if (options.negativeText.isNotEmpty()) {
      builder.setNegativeButton(options.negativeText, listener)
    }

    if (options.neutralText.isNotEmpty()) {
      builder.setNeutralButton(options.neutralText, listener)
    }

    if (options.defaultSelectedIndex != null) {
      selectedIndex = options.defaultSelectedIndex
    }

    builder.setSingleChoiceItems(options.items, options.defaultSelectedIndex ?: -1, this.listener)

    return builder.create()
  }

  internal inner class NUISingleChoiceListener(private val callback: Callback, private val reactContext: ReactApplicationContext) :
    DialogInterface.OnClickListener, DialogInterface.OnDismissListener {

    private var callbackConsumed = false

    override fun onClick(dialog: DialogInterface, which: Int) {
      if (which >= 0 ) {
        this@NUISingleChoice.selectedIndex = which
        return
      }

      if (!callbackConsumed) {
        if (reactContext.hasActiveReactInstance()) {
          when (which) {
            DialogInterface.BUTTON_POSITIVE -> callback.invoke("positive", this@NUISingleChoice.selectedIndex)
            DialogInterface.BUTTON_NEGATIVE -> callback.invoke("negative", this@NUISingleChoice.selectedIndex)
            DialogInterface.BUTTON_NEUTRAL -> callback.invoke("neutral", this@NUISingleChoice.selectedIndex)
          }
          callbackConsumed = true
        }
      }
    }

    override fun onDismiss(dialog: DialogInterface?) {
      if (!callbackConsumed) {
        if (reactContext.hasActiveReactInstance()) {
          callback.invoke("dismissed", this@NUISingleChoice.selectedIndex)
          callbackConsumed = true
        }
      }
    }
  }

}

